File: 백준/Gold/1976. 여행 가자/여행 가자.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(int(1e6))


def find(num):
    if num != cities[num]:
        cities[num] = find(cities[num])
    return cities[num]


def union(n1, n2):
    n1, n2 = find(n1), find(n2)

    if n1 == n2: return
    if n1 < n2: cities[n2] = n1
    else: cities[n1] = n2


n = int(input())
m = int(input())
cities = [i for i in range(n + 1)]

for i in range(1, n + 1):
    ipt = list(map(int, input().split()))
    for j in range(n):
        if ipt[j] == 1:
            union(i, j + 1)

travel = list(map(int, input().split()))
root = find(travel[0])

for city in travel:
    if find(city) != root:
        print('NO')
        break
else: print('YES')

File: 백준/Gold/1976. 여행 가자/여행 가자.cc
#include <iostream>
#include <vector>

using namespace std;

vector<int> parent;

int find(int p) {
  if (parent[p] != p) parent[p] = find(parent[p]);
  return parent[p];
}

void _union(int p, int q) {
  p = find(p);
  q = find(q);
  if (p == q) return;
  if (p < q)
    parent[q] = p;
  else
    parent[p] = q;
  return;
}

int main() {
  int n, m;

  scanf("%d\n%d", &n, &m);

  for (int i = 0; i < n + 1; i++) {
    parent.push_back(i);
  }

  for (int i = 1; i < n + 1; i++) {
    int target;

    for (int j = 1; j < n + 1; j++) {
      scanf("%d", &target);
      if (i != j && target == 1) _union(i, j);
    }
  }

  int city, ref;
  bool result = true;

  scanf("%d", &city);
  ref = find(city);

  for (int i = 0; i < m - 1; i++) {
    scanf(" %d", &city);
    if (find(city) != ref) {
      result = false;
    }
  }

  if (result) printf("YES\n");
  else printf("NO\n");

  return 0;
}
